core/src/test/scala/unit/kafka/admin/PreferredReplicaLeaderElectionCommandTest.scala

```diff
@@ -0,0 +1,336 @@
+    assertEquals(testPartitionPreferredLeader, getLeader(testPartition))
+    val jsonFile = toJsonFile(testPartitionAndAssignment.keySet)
+    try {
+      // Now do the election, even though the preferred replica is *already* the leader
+      PreferredReplicaLeaderElectionCommand.run(Array(
+        "--bootstrap-server", bootstrapServer(),
+        "--path-to-json-file", jsonFile.getAbsolutePath))
+      // Check the leader for the partition still is the preferred one
+      assertEquals(testPartitionPreferredLeader, getLeader(testPartition))
+    } finally {
+      jsonFile.delete()
+    }
+  }
+
+  /** What happens if the preferred replica is offline? */
+  @Test
+  def testWithOfflinePreferredReplica() {
+    createTestTopicAndCluster(testPartitionAndAssignment)
+    bounceServer(testPartitionPreferredLeader, testPartition)
+    // Check the leader for the partition is not the preferred one
+    val leader = getLeader(testPartition)
+    assertNotEquals(testPartitionPreferredLeader, leader)
+    // Now kill the preferred one
+    servers(testPartitionPreferredLeader).shutdown()
+    // Now try to elect the preferred one
+    val jsonFile = toJsonFile(testPartitionAndAssignment.keySet)
+    try {
+      PreferredReplicaLeaderElectionCommand.run(Array(
+        "--bootstrap-server", bootstrapServer(),
+        "--path-to-json-file", jsonFile.getAbsolutePath))
+      fail();
+    } catch {
+      case e: AdminCommandFailedException =>
+        assertEquals("1 preferred replica(s) could not be elected", e.getMessage)
+        val suppressed = e.getSuppressed()(0)
+        assertTrue(suppressed.isInstanceOf[LeaderNotAvailableException])
+        assertTrue(suppressed.getMessage, suppressed.getMessage.contains("Failed to elect leader for partition test-0 under strategy PreferredReplicaPartitionLeaderElectionStrategy"))
+        // Check we still have the same leader
+        assertEquals(leader, getLeader(testPartition))
+    } finally {
+      jsonFile.delete()
+    }
+  }
+
+  /** What happens if the controller gets killed just before an election? */
+  @Test
+  def testTimeout() {
+    createTestTopicAndCluster(testPartitionAndAssignment)
+    bounceServer(testPartitionPreferredLeader, testPartition)
+    // Check the leader for the partition is not the preferred one
+    val leader = getLeader(testPartition)
+    assertNotEquals(testPartitionPreferredLeader, leader)
+    // Now kill the controller just before we trigger the election
+    servers(getController().get.config.brokerId).shutdown()
+    val jsonFile = toJsonFile(testPartitionAndAssignment.keySet)
+    try {
+      PreferredReplicaLeaderElectionCommand.run(Array(
+        "--bootstrap-server", bootstrapServer(),
+        "--path-to-json-file", jsonFile.getAbsolutePath),
+        timeout = 10000)
+      fail();
+    } catch {
+      case e: AdminCommandFailedException =>
+        assertEquals("1 preferred replica(s) could not be elected", e.getMessage)
```

Comment:
Hmm, this really depends on whether the controller is the bootstrap server. If yes, the client will get a timeout exception. In other cases, the client could get a successful response. The former is not necessarily guaranteed.

Reply:
As mentioned earlier, this only works if the controller is the bootstrap server. Otherwise, by the time that PreferredReplicaLeaderElectionCommand runs, a new controller could have been elected.